### DISCUSSION TITLE
Is it possible to dump whole trace before its export?

### DISCUSSION THREAD

**Discussion by magicliang** [open-telemetry/opentelemetry-java-instrumentation#6122] [Q&A]
My question is I want to dump trace after it is done, I don't know when It is done, but I know before a trace is exported to remote server, it must be already done.
Is there any config or interface I can use so that I can make a double-write, so I can dump trace locally?

**Comment by laurit** [comment]
Trace is a collection of spans, it doesn't have an explicit end, but rather you could say that the trace is completed when no more spans are received for it. In practice when a complete trace is desired one could apply some timeouts. For example you could say that a trace is max 10min long or if a trace has not received new spans for more than a minute it is complete.

**Reply by magicliang** [reply]
> Trace is a collection of spans, it doesn't have an explicit end, but rather you could say that the trace is completed when no more spans are received for it. In practice when a complete trace is desired one could apply some timeouts. For example you could say that a trace is max 10min long or if a trace has not received new spans for more than a minute it is complete.

I think I miss the key point, you are right, trace is a global concept. Then my question becomes this: is it possible to make a double-write to dump to span locally.

**Reply by laurit** [reply]
See https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md you can set `otel.traces.exporter` to multiple exporters by specifying them as a comma separated list. One of them could just dump the span data, for example see https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/logging/src/main/java/io/opentelemetry/exporter/logging/LoggingSpanExporter.java

**Reply by magicliang** [reply]
> See https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md you can set `otel.traces.exporter` to multiple exporters by specifying them as a comma separated list. One of them could just dump the span data, for example see https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/logging/src/main/java/io/opentelemetry/exporter/logging/LoggingSpanExporter.java

Is it possible to intercept this logging exporter? may be we don't want to read span info from log.

**Reply by trask** [reply]
If I understand, I think you would probably be better off writing your own SpanExporter for this

**Reply by magicliang** [reply]
> If I understand, I think you would probably be better off writing your own SpanExporter for this

That makes sense. Thank you for pointing out that.
